src/uast.c

```diff
@@ -0,0 +1,115 @@
+
+#include "uast.h"
+#include "roles.h"
+
+#define GET(__NODE__, __PROP__, ...) \
+  (api->impl.__PROP__(__NODE__, ##__VA_ARGS__))
+
+static xmlNodePtr create_xml_node(node_api *api, void *node,
+                                  xmlNodePtr parent) {
+  const char *internal_type = GET(node, internal_type);
+  xmlNodePtr xmlNode = xmlNewNode(NULL, BAD_CAST(internal_type));
+  xmlNode->_private = node;
+  if (parent) {
+    xmlAddChild(parent, xmlNode);
+  }
+
+  // Token
+  const char *token = GET(node, token);
+  if (token) {
+    xmlNewProp(xmlNode, BAD_CAST("token"), BAD_CAST(token));
+  }
+
+  // Roles
+  int roles_size = GET(node, roles_size);
+  for (int i = 0; i < roles_size; i++) {
+    uint16_t role = GET(node, roles, i);
+    const char *role_name = role_name_for_id(role);
+    if (role_name != NULL) {
+      xmlNewProp(xmlNode, BAD_CAST(role_name), NULL);
+    }
+  }
+
+  // Recursivelly visit all children
+  int children_size = GET(node, children_size);
+  for (int i = 0; i < children_size; i++) {
+    void *child = GET(node, children, i);
+    create_xml_node(api, child, xmlNode);
+  }
+  return xmlNode;
+}
+
+static xmlDocPtr create_document(node_api *api, void *node) {
+  xmlDocPtr doc = xmlNewDoc(BAD_CAST("1.0"));
+  xmlNodePtr xmlNode = create_xml_node(api, node, NULL);
+  xmlDocSetRootElement(doc, xmlNode);
+  return doc;
+}
+
+//////////////////////////////
+//////// PUBLIC API /////////
+//////////////////////////////
+
+node_api *new_node_api(node_iface iface) {
+  xmlInitParser();
+  node_api *api = calloc(1, sizeof(node_api));
```

Comment:
* a `malloc` wouldn't require the first argument
* `#include <stdlib.h>`
* Error checking missing